File: note.py
from colorama import Fore


class Note:
    TYPES_TO_COLORS = {
        "static": Fore.WHITE,
        "null": Fore.GREEN,
        "temp": Fore.RED,
    }

    def __init__(self, value):
        self.value = value
        self.type = "null" if value == 0 else "static"

    def get_color(self):
        return Note.TYPES_TO_COLORS.get(self.type, Fore.RESET)

    def change_value(self, value):
        if self.value != 0 or value == 0:
            return
        self.value = value
        self.type = "temp"

    def reset(self):
        self.value = 0
        self.type = "null"

File: Models/Note.cs
namespace SudokuBruteForce.Models
{
    public class Note
    {
        public int value;
        public string type;

        public static readonly Dictionary<string, ConsoleColor> TypesToColors = new Dictionary<string, ConsoleColor>
        {
            {"static", ConsoleColor.White},
            {"null", ConsoleColor.Green},
            {"temp", ConsoleColor.Red},
            {"inserted", ConsoleColor.Blue}
        };

        public Note(int value)
        {
            this.value = value;
            this.type = value == 0 ? "null" : "static";
        }

        public ConsoleColor GetColor()
        {
            return Note.TypesToColors[this.type];
        }

        public void SetAsInserted()
        {
            this.type = "inserted";
        }

        public void ChangeValue(int value, int[] sums, int[] usedNotes)
        {
            if (this.value != 0 || value == 0 || usedNotes.Contains(value)) 
                return;

            this.value = value;

            if (sums.Any(sum => sum > Grid.MaxSum))
            {
                this.type = "null";
                this.value = 0;
                return;
            }

            if (sums.All(sum => sum == Grid.MaxSum))
            {
                this.type = "temp";
            }
            else if (sums.Any(sum => sum < Grid.MaxSum))
            {
                this.type = "temp";
            }
        }

        public void Reset()
        {
            this.type = "null";
            this.value = 0;
        }
    }
}

